src/main/java/com/renatusnetwork/momentum/data/menus/MenuItemAction.java

```diff
@@ -619,22 +619,31 @@ public static void performLevelTeleport(PlayerStats playerStats, Level level)
         StatsManager statsManager = Momentum.getStatsManager();
         player.closeInventory();
 
+        boolean enteringMasteryOfSameLevel = false;
+        boolean shiftClicked = Momentum.getMenuManager().containsShiftClicked(playerStats);
+
         if (nonLevelTeleportConditions(playerStats))
         {
-            if (level.playerHasRequiredLevels(playerStats))
-            {
+            if (level.playerHasRequiredLevels(playerStats)) {
                 // if the level has perm node, and player does not have perm node
-                if (level.hasPermissionNode() && !player.hasPermission(level.getRequiredPermission()))
-                {
+                if (level.hasPermissionNode() && !player.hasPermission(level.getRequiredPermission())) {
                     player.sendMessage(Utils.translate("&cYou do not have permission to enter this level"));
                     return;
                 }
 
-                // if player is in level and their level is the level they clicked on, cancel
-                if (playerStats.inLevel() && level.equals(playerStats.getLevel()))
-                {
-                    player.sendMessage(Utils.translate("&cUse the door to reset the level you are already in"));
-                    return;
+                // if theyre entering/leaving the same level from mastery or not
+                if (playerStats.inLevel() && level.equals(playerStats.getLevel())) {
+                    if (!playerStats.isAttemptingMastery()) {
+                        if (!shiftClicked || playerStats.hasMasteryCompletion(level)) {
+                            player.sendMessage(Utils.translate("&cUse the door to reset the level you are already in"));
+                            return;
+                        }
+                        enteringMasteryOfSameLevel = true;
+                    }
+                    else {
```

Comment:
this frustrates me, either } else { or
}
else
{

>:(

Reply:
changed